proposals/gc/MVP.md

```diff
@@ -0,0 +1,272 @@
+  - `fieldtype ::= <mutability> <storagetype>`
+  - `storagetype ::= <valtype> | <packedtype>`
+  - `packedtype ::= i8 | i16`
+
+* Unpacking a storage type yields `i32` for packed types, otherwise the type itself
+  - `unpacked(t) = t`
+  - `unpacked(pt) = i32`
+
+
+#### Imports
+
+* `type <typetype>` is an import description with an upper bound
+  - `importdesc ::= ... | type <reftype>`
+  - Note: `type` may get additional parameters in the future
+
+
+#### Exports
+
+* `type <typeidx>` is an export description
+  - `exportdesc ::= ... | type <typeidx>`
+  - `type $t ok` iff `$t` is defined in the context
+
+
+#### Subtyping
+
+Greatest fixpoint (co-inductive interpretation) of the given rules (implying reflexivity and transitivity).
+
+* `eqref` is a subtype of `anyref`
+  - `eqref <: anyref`
+  - Note: `int31ref` and `anyfunc` are *not* a subtypes of `eqref`, i.e., those types do not expose reference equality
+
+* `nullref` is a subtype of `eqref`
+  - `nullref <: eqref`
+  - Note: `nullref` is *not* a subtype of `anyfunc`, `int31ref`or any concrete reference type, i.e., those types are not nullable
+
+* `int31ref` is a subtype of `anyref`
+  - `int31ref <: anyref`
+
+* Any concrete reference type is a subtype of `anyref`
+  - `ref $t <: anyref`
+  - Note: concrete reference types are *not* a supertypes of `nullref`, i.e., nut nullable
+
+* Any function reference type is a subtype of `anyfunc`
+  - `ref $t <: anyfunc`
+     - iff `$t = <functype>`
+
+* Any concrete reference type is a subtype of `eqref` if its not a function
+  - `ref $t <: eqref`
+     - if `$t = <structtype>` or `$t = <arraytype>`
+     - or `$t = type rt` and `rt <: eqref`
+  - TODO: provide a way to make data types non-eq, especially immutable ones
+
+* Concrete reference types are covariant
+  - `ref $t1 <: ref $t2`
+     - iff `$t1 <: $t2`
+
+* Structure types support width and depth subtyping
+  - `struct <fieldtype1>* <fieldtype1'>* <: struct <fieldtype2>*`
+    - iff `(<fieldtype1> <: <fieldtype2>)*`
+
+* Array types support depth subtyping
+  - `array <fieldtype1> <: array <fieldtype2>`
+    - iff `<fieldtype1> <: <fieldtype2>`
+
+* Field types are covariant if they are immutable, invariant otherwise
+  - `const <valtype1> <: const <valtype2>`
+    - iff `<valtype1> <: <valtype2>`
+  - `var <valtype> <: var <valtype>`
+  - Note: mutable fields are *not* subtypes of immutable ones, so `const` really means constant, not read-only
+
+
+#### Defaultability
+
+* Any numeric value type is defaultable
+
+* A reference value type is defaultable if it is not of the form `ref $t`
+
+* Locals must have a type that is defaultable.
+
+
+### Instructions
+
+#### Equality
+
+* `ref.eq` compares two references whose types support equality
+  - `ref.eq : [eqref eqref] -> [i32]`
+
+
+#### Functions
+
+* `ref.func` creates a function reference from a function index
+  - `ref.func $f : [] -> (ref $t)`
+     - iff `$f : $t`
```

Comment:
I had been on the edge of whether function references were needed for the GC MVP, but the previous discussion about passing closures to and from JS without wrapping each time seemed to boost the priority.  Given that, could we also have the `bind` operator which can be used to manufacture closures?

Reply:
Oh, but closures are different from mere function references, since they have a different runtime representation. Hence they imply a new type, so I considered them post-MVP.

Reply:
Ah, I had assumed that they would have the same type and that they would also have the same representation.  That is, even without a bound dynamic environment, `ref.func` is still a closure because it entrains the `WebAssembly.Instance`.  We represent these as `JSFunction` objects internally in SM.  I had entertained the ideal of representing function references as raw function pointers, but I think this gets pretty hairy when GC and multi-instance linking enters the picture.

Reply:
I added it to the question section.